src/util/Timer.cpp

```diff
@@ -346,115 +323,72 @@ VirtualClock::crank(bool block)
     {
         return 0;
     }
-
-    size_t nWorkDone = 0;
-
+    size_t progressCount = 0;
     {
-        std::lock_guard<std::recursive_mutex> lock(mDelayExecutionMutex);
-        // Adding to mDelayedExecutionQueue is now restricted to main thread.
-
-        mDelayExecution = true;
-
+        std::lock_guard<std::recursive_mutex> lock(mDispatchingMutex);
+        mDispatching = true;
+        mLastDispatchStart = now();
         nRealTimerCancelEvents = 0;
-
         if (mMode == REAL_TIME)
         {
-            // Fire all pending timers.
-            // May add work to mDelayedExecutionQueue
-            nWorkDone += advanceToNow();
+            // Dispatch all pending timers.
+            progressCount += advanceToNow();
         }
 
-        // Pick up some work off the IO queue.
-        // Calling mIOContext.poll() here may introduce unbounded delays
-        // to trigger timers.
-        size_t lastPoll;
-        // to trigger timers so we instead put bounds on the amount of work we
-        // do at once both in size and time. As MAX_CRANK_WORK_DURATION is
-        // not enforced in virtual mode, we use a smaller batch size.
-        // We later process items queued up in the execution queue, that has
-        // similar parameters.
-        size_t const WORK_BATCH_SIZE = 100;
-        std::chrono::milliseconds const MAX_CRANK_WORK_DURATION{200};
-
-        YieldTimer ioYt(*this, MAX_CRANK_WORK_DURATION, WORK_BATCH_SIZE);
-        do
-        {
-            // May add work to mDelayedExecutionQueue.
-            lastPoll = mIOContext.poll_one();
-            nWorkDone += lastPoll;
-        } while (lastPoll != 0 && ioYt.shouldKeepGoing());
+        // Dispatch some IO event completions.
+        mLastDispatchStart = now();
+        size_t ioDivisor = mActionScheduler->isOverloaded() ? 2 : 1;
```

Comment:
not sure about the hardcoded "2" here: why not have this depend on how long we've been overloaded (can be exponential to the time spent overloaded) - that way we can pretty much guarantee to get unstuck. As IO can be tied to the number of connections for example, I don't know if "2" would be enough to stop the local node to be overloaded.

Reply:
Ok. Accepting the exponential-backoff you added. Looks fine.